docs/enhancements/multi-rhel-fips.md

```diff
@@ -0,0 +1,244 @@
+  concurrently (maybe we should create a metric for this), but the telco scale
+  team regularly runs several hundred concurrently maxing out at over three
+  thousand in a few hours. Unless we're cleaning up the jobs rather aggressively
+  I don't think it would be good to create this many.
+
+- Multiple jobs would need to be run on a single assets directory. This seems
+  prohibitively complex compared to the proposed solution. During a single
+  install the following installer commands are used:
+  - `openshift-baremetal-install create manifests`
+  - `openshift-baremetal-install create single-node-ignition-config` or
+    `openshift-baremetal-install create ignition-configs` (depending on HA mode)
+
+### Run the matching installer on the assisted-service container
+
+Clusters that have installers that already match the assisted service container's
+architecture could be handled by the assisted-service container as we do today.
+This would require one less image and container per pod, but having the same
+process for every cluster install would be easier to understand and maintain.
+
+### Use RPC over HTTP
+
+[Go's RPC](https://pkg.go.dev/net/rpc@go1.22.3) could be used instead of a direct
+HTTP server (RPC can be hosted over HTTP, but that's not what is being addressed
+here). RPC would make this a simpler change as the code for generating the
+manifests is already contained in a single package, but RPC would be a strange
+choice if we were to move the handling into a truly separate service in the
+future.
+
+### Install multiple libraries on the same image
+
+It may be possible to install both versions of the shared libraries required by
+the installers (libcrypto and libssl?) for FIPS compliance on a single image.
+This would require much less change and should be significantly quicker to
+implement, but it's not clear if these would be possible or supportable.
+This could be achieved by any of the following methods:
+
+1. Create a separate userspace for el8 libraries and chroot when those libraries
+   are required.
+  - This seems a bit complicated and it will likely make our image quite a bit
+    larger than it already is (~1.3G).
+2. Install both versions of the required packages simultaneously.
+  - Not sure if this is possible given that the packages share a name and are
+    only different in version.
+3. Use multi-layer container builds to copy the libraries from an el8 image to a
+   directory on the el9 image and use `LD_PRELOAD` or manipulate `LD_LIBRARY_PATH`
+   to point the el8 installer binaries to the correct libraries.
+
+The approach using chroot worked, but FIPS SMEs said that the container base
+image *must* match the installer for the resulting cluster to be considered
+FIPS-compliant so none of these multi-library options are valid.
+
+### Publish multiple assisted-service images
+
+It's likely that a user installing in FIPS mode will only be interested in
```

Comment:
Why do we think this is true?

Everyone starts with a single version. But existing clusters and their use cases lag behind new clusters and new use cases. I would expect that most users end up with a mix of openshift versions over time.

Keep in mind that users need the ability to re-provision their clusters at any time, for recovery from all kind of unexpected events. And a growing number of customers depend on regular "cluster as a service" provisioning. Telling such a user that they can only (re-)provision a subset of openshift versions would generally be a big disadvantage.

Even a large-scale user with relatively uniform versions is not going to upgrade all of their clusters at once. It would not be ideal to put them in a position where they can only provision clusters for the newest openshift versions in their fleet.

Is there reason to think that a FIPS user would have different patterns and expectations?

Reply:
You're right that this is an assumption on my part, but my experience with even moderately conservative customers is that they have a single version of OCP that is vetted and allowed in the org (at least in production).

My guess was that any user interested in full FIPS compliance would be even more strict than that, but I have no actual users to back any of this up.

Reply:
As for upgrade and reprovisioning. I do consider this option a temporary solution that buys us time to properly think through and vet something that would work more generically. Ideally in the version following the one this alternative would be implemented in we'd implement the main body of this enhancement (or something similar).

Reply:
They start with a single vetted version, then that progresses over time, but upgrades lag.

I think it's still better to implement this enhancement as described, which would fully preserve the existing use cases and user experience, and then optimize it later as needed.

Reply:
@mhrivnak all good. can we find such customer now? we'll have it as an rfe and improve in the next release. I dont think we should restructure the whole service 2 weeks before the release.
for next release, lets reconsider.